### PR TITLE
add lightgbm.booster support

### PR DESCRIPTION
**Pull request by qh582**
mainly followed the logic of xgboost.booster support

### CODE REVIEW

**Comment by lopuhin** [comment]
Thanks for the PR @qh582 , at first sight it looks good 👍 
I left a comment on how to fix the build. It would be also great if you could add tests for lightgbm.booster support, so that we don't break it in the future.

**Comment by qh582** [comment]
@lopuhin Thanks for review and suggestion. 
I am doing the `explain_prediction` for regression part test, and I got 'y=y' in stead of 'y' as shown in the screenshot:
image
This seems ok for me, but I found eli5 using `assert '<b>y</b>' in strip_blanks(expl_html)`  in `tests.test_sklearn_explain_prediction.assert_trained_linear_regression_explained` which point that it shoud be 'y' in my case. I am wondering if I missed the point here.

**Comment by kmike** [comment]
@qh582 looks good! Would you mind updating a few notes in docs, mentioning this new feature? 

* https://github.com/TeamHG-Memex/eli5/blob/master/README.rst
* https://github.com/TeamHG-Memex/eli5/blob/master/docs/source/libraries/lightgbm.rst
* https://github.com/TeamHG-Memex/eli5/blob/master/docs/source/overview.rst

**Comment by qh582** [comment]
@kmike I think this is ready, please check :)

File: docs/source/libraries/lightgbm.rst [hunk 1]
```diff
@@ -51,6 +50,14 @@ for ``lightgbm.LGBMClassifer`` and ``lightgbm.LGBMRegressor``:
   estimator. Set it to True if you're passing ``vec``,
   but ``doc`` is already vectorized.
 
+``lightgbm.Booster`` estimator accepts one more optional argument:
+
+* ``is_regression`` - True if solving a regression problem
+  ("objective" starts with "reg")
+  and False for a classification problem.
+  If not set, regression is assumed for a single target estimator
+  and proba will not be shown unless the ``target_names`` is defined as a list with length of two.
```
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]: This seems to contradict a note above (line 25, "target_names is ignored") - does the note about target_names apply only to classifier/regressor, but not for Booster?

File: docs/source/libraries/lightgbm.rst [hunk 2]
```diff
@@ -5,13 +5,13 @@ LightGBM
 
 LightGBM_ is a fast Gradient Boosting framework; it provides a Python
 interface. eli5 supports :func:`eli5.explain_weights`
-and :func:`eli5.explain_prediction` for ``lightgbm.LGBMClassifer``
-and ``lightgbm.LGBMRegressor`` estimators.
+and :func:`eli5.explain_prediction` for ``lightgbm.LGBMClassifer``, ``lightgbm.LGBMRegressor`` and ``lightgbm.Booster``  estimators. It is tested against LightGBM
+master branch.
```
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]: "It is tested against LightGBM master branch." is removed in https://github.com/TeamHG-Memex/eli5/commit/63e99182dc682bbf225355c80a24807396a747b6, sorry for the confusion. It shouldn't be re-added.

File: docs/source/libraries/lightgbm.rst (hunk 1 above)
Comment (qh582) [reply docs/source/libraries/lightgbm.rst]: for a single target estimater, booster cannot recognize wheather it is a regression problem or not. We assume it is regression in default, unless users set it as  a classification problem by assigning  'target names' input [0,1] etc. Only in this case  'target names' is used.  Should I remove " target names is ignored" to prevent confusion？

File: docs/source/libraries/lightgbm.rst (hunk 2 above)
Comment (qh582) [reply docs/source/libraries/lightgbm.rst]: got it, I will remove that line in next commit

File: docs/source/libraries/lightgbm.rst (hunk 1 above)
Comment (kmike) [reply docs/source/libraries/lightgbm.rst]: target names is still ignored for classifier/regressor, right? I think it makes sense to clarify it - just say that it is ignored for LGBMClassifier / LGBMRegressor, but used for lightgbm.Booster.

"is defined as a list with length of two" - sohuld it be 2 elements, or 2+ is also supported?

Comment (qh582) [reply docs/source/libraries/lightgbm.rst]: Well, for single target booster,  only 2 elements target_name point booster to classification. Otherwise, this booster is regression (there is risk user assign wrong number of elements like 3, but not sure how eli5 will raise error. 2+ should not support here).

**Comment by qh582** [comment]
@kmike Hope the new commit help. My previous piece of code is not strict enough.

**Comment by WestXu** [comment]
It's been two month, could someone tell me why this merge hasn't been closed already? I really need this.

**Comment by lopuhin** [comment]
I think PR is in great shape and we just need to check it again and merge it, hope to find time for it soonish, sorry for delay.

File: docs/source/libraries/lightgbm.rst [hunk 3]
```diff
@@ -5,13 +5,12 @@ LightGBM
...
 interface. eli5 supports :func:`eli5.explain_weights`
-and :func:`eli5.explain_prediction` for ``lightgbm.LGBMClassifer``
-and ``lightgbm.LGBMRegressor`` estimators.
+and :func:`eli5.explain_prediction` for ``lightgbm.LGBMClassifer``, ``lightgbm.LGBMRegressor`` and ``lightgbm.Booster``  estimators. 
 
 .. _LightGBM: https://github.com/Microsoft/LightGBM
 
 :func:`eli5.explain_weights` uses feature importances. Additional
-arguments for LGBMClassifier and LGBMClassifier:
+arguments for LGBMClassifier , LGBMClassifier and lightgbm.Booster:
```
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]: a nitpick: extra whitespace before comma

File: docs/source/libraries/lightgbm.rst [hunk 4]
```diff
@@ -22,7 +21,7 @@ arguments for LGBMClassifier and LGBMClassifier:
   - 'weight' - the same as 'split', for better compatibility with
     :ref:`library-xgboost`.
 
-``target_names`` and ``target`` arguments are ignored.
+``target_names`` arguement is ignored for ``lightgbm.LGBMClassifer`` / ``lightgbm.LGBMRegressor``, but used for ``lightgbm.Booster``. ``target`` argument is ignored.
```
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]:
```suggestion
``target_names`` argument is ignored for ``lightgbm.LGBMClassifer`` / ``lightgbm.LGBMRegressor``, but used for ``lightgbm.Booster``. ``target`` argument is ignored.
```

File: docs/source/libraries/lightgbm.rst (hunk 3 above)
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]:
```suggestion
arguments for LGBMClassifier, LGBMClassifier and lightgbm.Booster:
```

File: docs/source/libraries/lightgbm.rst [hunk 5]
```diff
@@ -22,7 +21,7 @@ arguments for LGBMClassifier and LGBMClassifier:
...
 
 .. note::
     Top-level :func:`eli5.explain_weights` calls are dispatched
```
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]: I think lightgbm.Booster should be mentioned here as well.

File: docs/source/libraries/lightgbm.rst [hunk 6]
```diff
@@ -50,6 +49,14 @@ for ``lightgbm.LGBMClassifer`` and ``lightgbm.LGBMRegressor``:
   estimator. Set it to True if you're passing ``vec``,
   but ``doc`` is already vectorized.
 
+``lightgbm.Booster`` estimator accepts one more optional argument:
+
+* ``is_regression`` - True if solving a regression problem
```
Comment (kmike) [inline docs/source/libraries/lightgbm.rst]: Is this parameter supported? It is not an argument of explain_prediction_lightgbm.

File: docs/source/overview.rst [hunk 7]
```diff
@@ -31,7 +31,7 @@ following machine learning frameworks and packages:
   of XGBClassifier, XGBRegressor and xgboost.Booster.
 
 * :ref:`library-lightgbm` - show feature importances and explain predictions
-  of LGBMClassifier and LGBMRegressor.
+  of LGBMClassifier , LGBMRegressor and lightgbm.Booster.
```
Comment (kmike) [inline docs/source/overview.rst]:
```suggestion
  of LGBMClassifier, LGBMRegressor and lightgbm.Booster.
```

File: eli5/lightgbm.py [hunk 8]
```diff
@@ -32,13 +32,15 @@ def explain_weights_lightgbm(lgb,
...
+    LGBMClassifier or LGBMRegressor, or via lightgbm.Booster) as feature importances.
 
     See :func:`eli5.explain_weights` for description of
     ``top``, ``feature_names``,
     ``feature_re`` and ``feature_filter`` parameters.
 
-    ``target_names`` and ``targets`` parameters are ignored.
+    ``target_names`` arguement is ignored for ``lightgbm.LGBMClassifer`` / ``lightgbm.LGBMRegressor``, 
+    but used for ``lightgbm.Booster``. 
+    ``target`` argument is ignored.
```
Comment (kmike) [inline eli5/lightgbm.py]:
```suggestion
    ``targets`` argument is ignored.
```

**Comment by Dronablo** [comment]
Any plans about merging it, @lopuhin?

**Comment by lopuhin** [comment]
Thanks you! This were merged in eli5-org/eli5#7 and released to PyPI with v0.11
